crates/gpui/src/platform/windows/dispatcher.rs

```diff
@@ -1,74 +1,119 @@
 use parking::Parker;
 use parking_lot::Mutex;
-use windows::Win32::{Foundation::*, System::Threading::*};
+use util::ResultExt;
+use windows::{
+    Foundation::TimeSpan,
+    System::{
+        DispatcherQueue, DispatcherQueueController, DispatcherQueueHandler,
+        Threading::{
+            ThreadPool, ThreadPoolTimer, TimerElapsedHandler, WorkItemHandler, WorkItemOptions,
+            WorkItemPriority,
+        },
+    },
+    Win32::System::WinRT::{
+        CreateDispatcherQueueController, DispatcherQueueOptions, DQTAT_COM_NONE,
+        DQTYPE_THREAD_CURRENT,
+    },
+};
 
 use crate::{PlatformDispatcher, TaskLabel};
 
+macro_rules! generate_handler {
+    ($handler_type: tt, $runnable: ident, $has_param: ident) => {{
+        macro_rules! generate_func {
+            (true, $function: block) => {
+                $handler_type::new(move |_| $function)
+            };
+            (false, $function: block) => {
+                $handler_type::new(move || $function)
+            };
+        }
+        let runnable = $runnable;
+        let task_wrapper = TaskWrapper(runnable.into_raw().as_ptr() as *mut c_void);
+        generate_func!($has_param, {
+            let task = unsafe {
+                let captured = task_wrapper;
+                Runnable::<()>::from_raw(std::ptr::NonNull::new_unchecked(captured.0 as *mut ()))
+            };
+            task.run();
+            Ok(())
+        })
+    }};
+}
+
 pub(crate) struct WindowsDispatcher {
-    threadpool: PTP_POOL,
-    main_sender: Sender<Runnable>,
+    controller: DispatcherQueueController,
+    main_queue: DispatcherQueue,
     parker: Mutex<Parker>,
     main_thread_id: ThreadId,
-    dispatch_event: HANDLE,
 }
 
+unsafe impl Send for WindowsDispatcher {}
+unsafe impl Sync for WindowsDispatcher {}
+
 impl WindowsDispatcher {
-    pub(crate) fn new(main_sender: Sender<Runnable>, dispatch_event: HANDLE) -> Self {
-        let parker = Mutex::new(Parker::new());
-        let threadpool = unsafe {
-            let ret = CreateThreadpool(None);
-            if ret.0 == 0 {
-                panic!(
-                    "unable to initialize a thread pool: {}",
-                    std::io::Error::last_os_error()
-                );
-            }
-            // set minimum 1 thread in threadpool
-            let _ = SetThreadpoolThreadMinimum(ret, 1)
-                .inspect_err(|_| log::error!("unable to configure thread pool"));
-
-            ret
+    pub(crate) fn new() -> Self {
+        // Windows 10 Fall Creators Update (introduced in 10.0.16299.0)
+        let controller = unsafe {
+            let options = DispatcherQueueOptions {
+                dwSize: std::mem::size_of::<DispatcherQueueOptions>() as u32,
+                threadType: DQTYPE_THREAD_CURRENT,
+                apartmentType: DQTAT_COM_NONE,
+            };
+            CreateDispatcherQueueController(options).unwrap()
         };
+        let main_queue = controller.DispatcherQueue().unwrap();
```

Comment:
I'm noticing that we're using the same queue for everything here, and that we're only calling `TryEnqueue`....

It seems like all of this is making Zed single threaded on windows? That seems wrong.

Reply:
On windows, there is one default thread pool per process, we use this default thread pool here. In the previous implementation, because the default thread pool could have a minimum of 0 threads (or workers), and the default thread pool could not be configured, so I created a new thread pool with a minimum of 1 worker. However, this new API does not provide a similar functionality, os I just use the default thread pool here.

Reply:
Ok, I see how it all comes together now. Thanks for letting me know :)